### PR TITLE
Tic Tac Toe Game Complete feature

### PR DESCRIPTION
**Pull request by osinakayah**

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            return
+        }
+        const playerOne = PlayerFactory(playerOneName, 'X')
+        const playerTwo = PlayerFactory(playerTwoName, 'O')
+        GameBoard.init('game_board', playerOne, playerTwo);
+
+        document.getElementById('player_one_name_display').innerHTML = playerOneName
+        document.getElementById('player_two_name_display').innerHTML = playerTwoName
+      })
+    </script>
```
Comment (heyapricot) [inline index.html]: What about storing this in another script. Maybe `main.js`?

File: js/script.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            case (gameboard[1] === 'O' && gameboard[4] === 'O' && gameboard[7] === 'O') || (gameboard[1] === 'X' && gameboard[4] === 'X' && gameboard[7] === 'X'):
+                return true
+            case (gameboard[2] === 'O' && gameboard[5] === 'O' && gameboard[8] === 'O') || (gameboard[2] === 'X' && gameboard[5] === 'X' && gameboard[8] === 'X'):
+                return true
+            case (gameboard[0] === 'O' && gameboard[4] === 'O' && gameboard[8] === 'O') || (gameboard[0] === 'X' && gameboard[4] === 'X' && gameboard[8] === 'X'):
+                return true
+            case (gameboard[2] === 'O' && gameboard[4] === 'O' && gameboard[6] === 'O') || (gameboard[2] === 'X' && gameboard[4] === 'X' && gameboard[6] === 'X'):
+                return true
+        }
+    }
```
Comment (heyapricot) [inline js/script.js:91]: I don't like how heavily hardcoded this is. It checks specifically for 'X' or 'O'. Also, the positions are hardcoded. 
Can we:
- Build a method that checks if there is any 3 element sequence?
- Build one or multiple methods that check if there is a sequence in a row, column or diagonals?

File: js/script.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    const isTie = (gameboard) => gameboard.filter(gameSpot => gameSpot === null).length === 0
```
Comment (heyapricot) [inline js/script.js:92]: Do we need to filter and generate a new array?
The way I see it the logic for this project works like this:
Is the board full?
Yes: The game is over. No: Game continues.
After checking if the board is full
Is there a sequence? There is a winner else is a draw

File: js/script.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+      }
+      renderGameBoard(newGameboardDrawing)
+    }
+
+    const attachEvent = (gameboard) => {
+      if (!hasGameStarted) {
+        return
+      }
+      const cells = document.getElementsByClassName('tic-tac-toe-cell');
+      for (let i = 0; i < cells.length; i ++) {
```
Comment (heyapricot) [inline js/script.js:125]: You can use [forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) instead

File: js/script.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        playerOne.setPlayerTurn(false);
+        playerTwo.setPlayerTurn(false)
+        hasGameStarted = false
+      }
+      if (isTie(newGameboardDrawing)) {
+        hasGameStarted = false
+        alert('Is a tie')
+      }
+      renderGameBoard(newGameboardDrawing)
+    }
```
Comment (heyapricot) [inline js/script.js:117]: This seems like a lot. You generate a new board array with the latest move and check if that move is going to win or if it's a tie. Why did you go with this implementation?

File: js/script.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        }
+      }
+    }
+    const renderGameBoard = (gameboard) => {
+        const view = gameboard.reduce((sumPositions, currentPosition, index) => {
+            return `${sumPositions}
+                <div data-cell-index="${index}" class="col-4 tic-tac-toe-cell">${currentPosition === null ? '' : currentPosition}</div>
+            `
+        }, '');
+        document.getElementById(gameBoardViewContainer).innerHTML = view;
```
Comment (heyapricot) [inline js/script.js:148]: [I favor the use of document.createElement over the use of innerHTML](https://stackoverflow.com/questions/2946656/advantages-of-createelement-over-innerhtml)

File: js/script.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+            playerOne = _playerOne;
+            playerTwo = _playerTwo;
+
+            playerOne.setPlayerTurn(true)
+            playerTwo.setPlayerTurn(false)
+
+            renderGameBoard(Array(9).fill(null));
+
+        }
```
Comment (heyapricot) [inline js/script.js:166]: Why is this function defined directly in the return?

**Review by heyapricot — CHANGES_REQUESTED** [review]
# It works! But it's a bit tangled 🍝
👋🏼 Hello! You got your JS Tic Tac Toe working. Yay! However, I think there is room for improvement in the project. I've added some annotations and raised an issue in the repository. Please address all of them and submit a new code review request. (You create a pull request for solving specifically the raised issue and submit a code review request for that specifically).

If you have any questions feel free to reach over crisp. Thanks!

File: js/script.js (hunk 5 above)
Comment (osinakayah) [reply js/script.js:117]: We are trying to follow the functional programming paradigm. So we didnt want the function to mutate the data, but rather create a new data. What is why we were using filter and map function a lot.

File: js/script.js (hunk 3 above)
Comment (osinakayah) [reply js/script.js:92]: Yes. We used filter() because we are following the functional programming paradigm.

File: js/script.js (hunk 7 above)
Comment (osinakayah) [reply js/script.js:166]: it renders the first empty board

**Review by heyapricot — CHANGES_REQUESTED** [review]
# Please take a look at the way you check for the sequence
Hello team! I was playing with the app and got an unexpected win. Please take a look at this issue and once solved submit a new code review request. Thanks!
Screen Shot 2019-06-03 at 11 59 14 AM
Screen Shot 2019-06-03 at 11 59 23 AM

**Comment by dipto0321** [comment]
Hi @osinakayah and @sanelca ,

Still, this application is not working properly. I am also having the same problem like @heyapricot Check out this picture.



How come `Dake` win the game?

So please fix that problem and submit another request to review this project. 

Thanks

**Comment by dipto0321** [comment]
Hi @osinakayah and @sanelca ,

This project is working now without any issue. Good Job. This is from me 👍 .

Thanks 🙏

**Pull request merged**